File: scala-3.6/tax-system/src/main/scala/com/rbleggi/taxsystem/specification/Specification.scala
package com.rbleggi.taxsystem.specification

import com.rbleggi.taxsystem.model.{State, TaxRule, Product}

trait Specification[T]:
  def matches(candidate: T): Boolean

  def and(other: Specification[T]): Specification[T] =
    (candidate: T) => this.matches(candidate) && other.matches(candidate)

  def or(other: Specification[T]): Specification[T] =
    (candidate: T) => this.matches(candidate) || other.matches(candidate)

class ProductSpecification(expectedProduct: Product) extends Specification[TaxRule]:
  override def matches(rule: TaxRule): Boolean = rule.product == expectedProduct

class StateSpecification(expectedState: State) extends Specification[TaxRule]:
  override def matches(rule: TaxRule): Boolean = rule.state == expectedState

class YearSpecification(expectedYear: Int) extends Specification[TaxRule]:
  override def matches(rule: TaxRule): Boolean = rule.year == expectedYear

File: kotlin/tax-system/src/main/kotlin/com/rbleggi/taxsystem/specification/Specification.kt
package com.rbleggi.taxsystem.specification

import com.rbleggi.taxsystem.model.Product
import com.rbleggi.taxsystem.model.State
import com.rbleggi.taxsystem.model.TaxRule

fun interface Specification<T> {
    fun matches(candidate: T): Boolean

    fun and(other: Specification<T>): Specification<T> =
        Specification { candidate -> this.matches(candidate) && other.matches(candidate) }

    fun or(other: Specification<T>): Specification<T> =
        Specification { candidate -> this.matches(candidate) || other.matches(candidate) }
}

class ProductSpecification(private val product: Product) : Specification<TaxRule> {
    override fun matches(candidate: TaxRule): Boolean = candidate.product == product
}

class StateSpecification(private val state: State) : Specification<TaxRule> {
    override fun matches(candidate: TaxRule): Boolean =
        candidate.state == state
}

class YearSpecification(private val year: Int) : Specification<TaxRule> {
    override fun matches(candidate: TaxRule): Boolean =
        candidate.year == year
}